src/main/java/org/sagebionetworks/bridge/services/UploadService.java

```diff
@@ -527,6 +566,89 @@ public void uploadComplete(String appId, UploadCompletionClient completedBy, Upl
         // Save uploadedOn date and uploadId to related adherence records.
         updateAdherenceWithUploadInfo(appId, upload);
     }
+
+    public void redriveUpload( byte[] fileBytes) throws IOException {
+        // Convert the byte[] array to a String
+        String fileContent = new String(fileBytes);
+
+        // Split the content into lines (each line is an entry - uploadId)
+        String[] lines = fileContent.split("\\r?\\n");
+
+        // Count the number of entries (lines)
+        int numberOfEntries = lines.length;
+
+        // Redrive upload.
+        if (numberOfEntries <= 10) {
+            redriveSmallAmountOfUploads(fileBytes);
+        } else {
+            redriveLargeAmountOfUploads(fileBytes);
+        }
+    }
+
+    private void redriveSmallAmountOfUploads(byte[] fileBytes) throws JsonProcessingException {
+        // Convert the byte[] array to a String
+        String fileContent = new String(fileBytes);
+
+        // Split the String into an array of lines
+        String[] linesArray = fileContent.split("\n");
+
+        // Convert the array of lines into a List
+        List<String> uploadIds = new ArrayList<>(Arrays.asList(linesArray));
+
+        for (String uploadId : uploadIds) {
+            Upload upload = getUpload(uploadId);
+            String appId = upload.getAppId();
+            if (appId == null) {
+                appId = healthCodeDao.getAppId(upload.getHealthCode());
+            }
+            uploadComplete(appId, UploadCompletionClient.REDRIVE, upload, true);
+            UploadValidationStatus validationStatus = pollUploadValidationStatusUntilComplete(uploadId);
+            if (validationStatus.getStatus() != UploadStatus.SUCCEEDED) {
+                logger.error("Redrive failed for uploadId=" + uploadId + ": " + COMMA_SPACE_JOINER.join(
+                        validationStatus.getMessageList()));
+            }
+        }
+    }
+
+    private void redriveLargeAmountOfUploads(byte[] fileBytes) throws IOException {
+        // set up s3 Key.
+        String currentYear = String.valueOf(DateUtils.getCurrentDateTime().getYear());
+        String currentMonth = String.valueOf(DateUtils.getCurrentDateTime().getMonthOfYear());
+        String s3Key = REDRIVE_UPLOAD_S3_KEY_PREFIX + currentYear + "-" + currentMonth;
+
+        // Upload file to S3 bucket
+        InputStream inputStream = new ByteArrayInputStream(fileBytes);
+        ObjectMetadata metadata = new ObjectMetadata();
+        metadata.setContentType(URLConnection.guessContentTypeFromStream(inputStream));
+
+        PutObjectRequest request = new PutObjectRequest(CONFIG_KEY_BACKFILL_BUCKET, s3Key, inputStream, metadata);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: This code is uploading an object of type stream to S3 without setting the content length of the object. Failure to specify a content length will cause the entire contents of the input stream to be buffered locally in memory so that the content length can be calculated, which can result in negative performance problems.

**Fix**: Specify the content length for the data stream in the object metadata parameter.

[Learn more](https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/s3/model/PutObjectRequest.html#PutObjectRequest-java.lang.String-java.lang.String-java.io.InputStream-com.amazonaws.services.s3.model.ObjectMetadata-)